[Bug]: Controllers pods going to crashloopback with panic error related to policiesCfg

### Version

5.4.1

### What Kubernetes platforms are you running on?

Other

### Steps to reproduce

apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    nginx.org/policies: emp-id-cdc-api-dev-uhg-com-mtls-policy  (ingressMTLs)
    name: emp-id-cdc-api-dev-uhg-com
spec:
  ingressClassName: f5-nginx-oss
  rules:
  - host: example.san.com
    http:
      paths:
      - backend:
          service:
            name: emp-id-cdc-app
            port:
              number: 80
        path: /
        pathType: Prefix

On multi-tenant cluster where 5.4.1 ingress nginx OSS controller running , customer  created ingressMTLS policy resource  and added to ingress object instead of virtual server  ( customer not aware it is not yet allowed in 5.4.1 ) which caused panic on all ingress controller pods and cluster traffic went down for all other applications.

Instead of rejecting the configuration all controller pods went to crashloopback and caused down time.

Error:

E0506 05:31:43.583564       1 panic.go:336] "Observed a panic" panic="runtime error: invalid memory address or nil pointer dereference" panicGoValue="\"invalid memory address or nil pointer dereference\"" stacktrace=<
	goroutine 902587 [running]:
	k8s.io/apimachinery/pkg/util/runtime.logPanic({0x28fdef0, 0x1bc4b81789c0}, {0x233cfa0, 0x42587d0})
		k8s.io/apimachinery@v0.35.3/pkg/util/runtime/runtime.go:132 +0xbc
	k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x28fe7c8, 0x1bc42aef9180}, {0x233cfa0, 0x42587d0}, {0x0, 0x0, 0x450000?})
		k8s.io/apimachinery@v0.35.3/pkg/util/runtime/runtime.go:107 +0x116
	k8s.io/apimachinery/pkg/util/runtime.HandleCrashWithContext({0x28fe7c8, 0x1bc42aef9180}, {0x0, 0x0, 0x0})
		k8s.io/apimachinery@v0.35.3/pkg/util/runtime/runtime.go:78 +0x55
	panic({0x233cfa0?, 0x42587d0?})
		runtime/panic.go:860 +0x13a
	github.com/nginx/kubernetes-ingress/internal/configs.(*policiesCfg).addIngressMTLSConfig(0x1bc4b7264db8, 0x1bc4b54ea980, {0x1bc4b81689c0, 0x35}, {0x1bc479b9cca0, 0xe}, {0x27836cd, 0x4}, 0x1, 0x1bc4b81786c0)
		github.com/nginx/kubernetes-ingress/internal/configs/policy.go:325 +0x3c8
	github.com/nginx/kubernetes-ingress/internal/configs.generatePolicies({_, _}, {{0x28e5a00, 0x1bc46b2e8160}, {0x1bc4807f3320, 0x1a}, {0x1bc479b9cca0, 0xe}, {0x1bc479b9cca0, 0xe}, ...}, ...)
		github.com/nginx/kubernetes-ingress/internal/configs/policy.go:1015 +0x5c9
	github.com/nginx/kubernetes-ingress/internal/configs.generateNginxCfg({0x1bc416a8c480, 0x1bc4b813a240, 0x0, 0x1bc4b8178900, 0x0, 0x1bc415f8c808, 0x0, 0x0, 0x0, 0x0, ...})
		github.com/nginx/kubernetes-ingress/internal/configs/ingress.go:170 +0xf30
	github.com/nginx/kubernetes-ingress/internal/configs.(*Configurator).addOrUpdateIngress(0x1bc4160ba210, 0x1bc4b813a240)
		github.com/nginx/kubernetes-ingress/internal/configs/configurator.go:440 +0x3a5
	github.com/nginx/kubernetes-ingress/internal/configs.(*Configurator).AddOrUpdateIngress(0x1bc4160ba210, 0x1bc4b813a240)
		github.com/nginx/kubernetes-ingress/internal/configs/configurator.go:302 +0x2f
	github.com/nginx/kubernetes-ingress/internal/k8s.(*LoadBalancerController).processChanges(0x1bc41589cf08, {0x1bc4b8178660, 0x1, 0x1bc4b7267ce8?})
		github.com/nginx/kubernetes-ingress/internal/k8s/controller.go:1317 +0x37a
	github.com/nginx/kubernetes-ingress/internal/k8s.(*LoadBalancerController).syncIngress(0x1bc41589cf08, {0x278c075?, {0x1bc448f36210?, 0x1bc4b7267d68?}})
		github.com/nginx/kubernetes-ingress/internal/k8s/controller.go:1734 +0x1d3
	github.com/nginx/kubernetes-ingress/internal/k8s.(*LoadBalancerController).sync(0x1bc41589cf08, {0x10?, {0x1bc448f36210?, 0x1bc4b7267d01?}})
		github.com/nginx/kubernetes-ingress/internal/k8s/controller.go:1050 +0x257
	github.com/nginx/kubernetes-ingress/internal/k8s.(*taskQueue).worker(0x1bc416100080)
		github.com/nginx/kubernetes-ingress/internal/k8s/task_queue.go:99 +0x95
	k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1({0x8ed218?, 0x1bc4a817db00?})
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:233 +0x13
	k8s.io/apimachinery/pkg/util/wait.BackoffUntilWithContext.func1({0x28fe7c8?, 0x1bc42aef9180?}, 0x4248b4?)
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:255 +0x51
	k8s.io/apimachinery/pkg/util/wait.BackoffUntilWithContext({0x28fe7c8, 0x1bc42aef9180}, 0x1bc4b7267f40, {0x28db4a0, 0x1bc4b5221e00}, 0x1)
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:256 +0xe5
	k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x8ed780?, {0x28db4a0?, 0x1bc4b5221e00?}, 0x0?, 0x1bc49b87bfa8?)
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:233 +0x46
	k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x1bc49b87bfc0, 0x3b9aca00, 0x0, 0x1, 0x1bc42aef9180)
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:210 +0x7f
	k8s.io/apimachinery/pkg/util/wait.Until(...)
		k8s.io/apimachinery@v0.35.3/pkg/util/wait/backoff.go:163
	github.com/nginx/kubernetes-ingress/internal/k8s.(*taskQueue).Run(...)
		github.com/nginx/kubernetes-ingress/internal/k8s/task_queue.go:48
	created by github.com/nginx/kubernetes-ingress/internal/k8s.(*LoadBalancerController).Run in goroutine 1
		github.com/nginx/kubernetes-ingress/internal/k8s/controller.go:685 +0xc1d
>
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x199c408]

--- a/internal/configs/policy.go
+++ b/internal/configs/policy.go
@@ -103,6 +103,14 @@ func newPoliciesConfig(bv bundleValidator) *policiesCfg {
 	}
 }
 
+// IsPolicySupportedOnIngress reports whether the given policy type is supported
+// on Ingress resources. Only AccessControl and CORS policies are supported on Ingress.
+// This is the single source of truth for the Ingress policy allowlist and must be kept
+// in sync with any callers that filter policies for Ingress resources (e.g. syncPolicy).
+func IsPolicySupportedOnIngress(pol *conf_v1.Policy) bool {
+	return pol.Spec.AccessControl != nil || pol.Spec.CORS != nil
+}
+
 func (p *policiesCfg) addAccessControlConfig(accessControl *conf_v1.AccessControl) *validationResults {
 	res := newValidationResults()
 	p.Allow = append(p.Allow, accessControl.Allow...)
@@ -445,7 +453,7 @@ func (p *policiesCfg) addIngressMTLSConfig(
 	secretKey := fmt.Sprintf("%v/%v", polNamespace, ingressMTLS.ClientCertSecret)
 	secretRef := secretRefs[secretKey]
 	if secretRef == nil {
-		res.addWarningf("IngressMTLS policy %q references an invalid secret %s: secret doesn't exist", polKey, secretKey)
+		res.addWarningf("IngressMTLS policy %s references a secret %s that is not available", polKey, secretKey)
 		res.isError = true
 		return res
 	}
@@ -1137,6 +1145,14 @@ func generatePolicies(
 		key := fmt.Sprintf("%s/%s", polNamespace, p.Name)
 
 		if pol, exists := policies[key]; exists {
+			// Reject policy types that are not supported on Ingress resources.
+			// IsPolicySupportedOnIngress is the single source of truth for the allowlist.
+			if ownerDetails.parentType == "ing" && !IsPolicySupportedOnIngress(pol) {
+				warnings.AddWarningf(ownerDetails.owner, "Policy %s is not supported on Ingress resources", key)
+				return policiesCfg{
+					ErrorReturn: &version2.Return{Code: 500},
+				}, warnings
+			}
 			var res *validationResults
 			switch {
 			case pol.Spec.AccessControl != nil:

--- a/internal/configs/policy.go
+++ b/internal/configs/policy.go
@@ -108,7 +108,12 @@ func newPoliciesConfig(bv bundleValidator) *policiesCfg {
 // This is the single source of truth for the Ingress policy allowlist and must be kept
 // in sync with any callers that filter policies for Ingress resources (e.g. syncPolicy).
 func IsPolicySupportedOnIngress(pol *conf_v1.Policy) bool {
-	return pol.Spec.AccessControl != nil || pol.Spec.CORS != nil
+	return pol.Spec.AccessControl != nil ||
+		pol.Spec.CORS != nil ||
+		pol.Spec.ExternalAuth != nil ||
+		pol.Spec.IngressMTLS != nil ||
+		pol.Spec.EgressMTLS != nil ||
+		pol.Spec.WAF != nil
 }
 
 func (p *policiesCfg) addAccessControlConfig(accessControl *conf_v1.AccessControl) *validationResults {

--- a/internal/configs/policy.go
+++ b/internal/configs/policy.go
@@ -107,6 +107,9 @@ func newPoliciesConfig(bv bundleValidator) *policiesCfg {
 // on Ingress resources. Only AccessControl and CORS policies are supported on Ingress.
 // This is the single source of truth for the Ingress policy allowlist and must be kept
 // in sync with any callers that filter policies for Ingress resources (e.g. syncPolicy).
+// To add support for a new policy type on Ingress, add its Spec field to this function.
+// Also ensure that createIngressEx() in controller.go loads any required secret or service
+// references for that policy type.
 func IsPolicySupportedOnIngress(pol *conf_v1.Policy) bool {
 	return pol.Spec.AccessControl != nil ||
 		pol.Spec.CORS != nil ||
@@ -1152,6 +1155,8 @@ func generatePolicies(
 		if pol, exists := policies[key]; exists {
 			// Reject policy types that are not supported on Ingress resources.
 			// IsPolicySupportedOnIngress is the single source of truth for the allowlist.
+			// If a new resource type with a subset of supported policy types is added,
+			// a matching parentType check must also be added in syncPolicy() in internal/k8s/policy.go.
 			if ownerDetails.parentType == "ing" && !IsPolicySupportedOnIngress(pol) {
 				warnings.AddWarningf(ownerDetails.owner, "Policy %s is not supported on Ingress resources", key)
 				return policiesCfg{

--- a/internal/configs/policy.go
+++ b/internal/configs/policy.go
@@ -458,7 +458,7 @@ func (p *policiesCfg) addIngressMTLSConfig(
 	secretKey := fmt.Sprintf("%v/%v", polNamespace, ingressMTLS.ClientCertSecret)
 	secretRef := secretRefs[secretKey]
 	if secretRef == nil {
-		res.addWarningf("IngressMTLS policy %s references a secret %s that is not available", polKey, secretKey)
+		res.addWarningf("IngressMTLS policy %q references a non-existent secret %s", polKey, secretKey)
 		res.isError = true
 		return res
 	}

--- a/internal/configs/policy.go
+++ b/internal/configs/policy.go
@@ -104,7 +104,7 @@ func newPoliciesConfig(bv bundleValidator) *policiesCfg {
 }
 
 // IsPolicySupportedOnIngress reports whether the given policy type is supported
-// on Ingress resources. Only AccessControl and CORS policies are supported on Ingress.
+// on Ingress resources.
 // This is the single source of truth for the Ingress policy allowlist and must be kept
 // in sync with any callers that filter policies for Ingress resources (e.g. syncPolicy).
 // To add support for a new policy type on Ingress, add its Spec field to this function.

--- a/internal/configs/policy_test.go
+++ b/internal/configs/policy_test.go
@@ -4242,14 +4242,14 @@ func TestIsPolicySupportedOnIngress(t *testing.T) {
 			expected: true,
 		},
 		{
-			name:     "IngressMTLS is not supported",
+			name:     "IngressMTLS is supported",
 			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{IngressMTLS: &conf_v1.IngressMTLS{ClientCertSecret: "ca"}}},
-			expected: false,
+			expected: true,
 		},
 		{
-			name:     "EgressMTLS is not supported",
+			name:     "EgressMTLS is supported",
 			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{EgressMTLS: &conf_v1.EgressMTLS{}}},
-			expected: false,
+			expected: true,
 		},
 		{
 			name:     "RateLimit is not supported",
@@ -4277,9 +4277,9 @@ func TestIsPolicySupportedOnIngress(t *testing.T) {
 			expected: false,
 		},
 		{
-			name:     "WAF is not supported",
+			name:     "WAF is supported",
 			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{WAF: &conf_v1.WAF{}}},
-			expected: false,
+			expected: true,
 		},
 		{
 			name:     "Cache is not supported",
@@ -4311,15 +4311,12 @@ func TestGeneratePolicies_UnsupportedOnIngress(t *testing.T) {
 	policyRef := []conf_v1.PolicyReference{{Name: "test-policy", Namespace: "default"}}
 
 	unsupportedPolicies := map[string]*conf_v1.Policy{
-		"IngressMTLS": {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{IngressMTLS: &conf_v1.IngressMTLS{ClientCertSecret: "ca"}}},
-		"EgressMTLS":  {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{EgressMTLS: &conf_v1.EgressMTLS{}}},
-		"RateLimit":   {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{RateLimit: &conf_v1.RateLimit{}}},
-		"JWTAuth":     {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{JWTAuth: &conf_v1.JWTAuth{}}},
-		"BasicAuth":   {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{BasicAuth: &conf_v1.BasicAuth{}}},
-		"OIDC":        {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{OIDC: &conf_v1.OIDC{}}},
-		"APIKey":      {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{APIKey: &conf_v1.APIKey{}}},
-		"WAF":         {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{WAF: &conf_v1.WAF{}}},
-		"Cache":       {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{Cache: &conf_v1.Cache{}}},
+		"RateLimit": {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{RateLimit: &conf_v1.RateLimit{}}},
+		"JWTAuth":   {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{JWTAuth: &conf_v1.JWTAuth{}}},
+		"BasicAuth": {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{BasicAuth: &conf_v1.BasicAuth{}}},
+		"OIDC":      {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{OIDC: &conf_v1.OIDC{}}},
+		"APIKey":    {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{APIKey: &conf_v1.APIKey{}}},
+		"Cache":     {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{Cache: &conf_v1.Cache{}}},
 	}
 
 	for policyType, pol := range unsupportedPolicies {

--- a/internal/configs/policy_test.go
+++ b/internal/configs/policy_test.go
@@ -3311,7 +3311,7 @@ func TestGeneratePoliciesFails(t *testing.T) {
 			},
 			expectedWarnings: Warnings{
 				nil: {
-					`IngressMTLS policy "default/ingress-mtls-policy" references an invalid secret default/ingress-mtls-secret: secret doesn't exist`,
+					`IngressMTLS policy "default/ingress-mtls-policy" references a non-existent secret default/ingress-mtls-secret`,
 				},
 			},
 			msg: "ingress mtls absent secret ref",

--- a/internal/configs/policy_test.go
+++ b/internal/configs/policy_test.go
@@ -4224,6 +4224,120 @@ func TestGeneratePoliciesFails(t *testing.T) {
 	}
 }
 
+func TestIsPolicySupportedOnIngress(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name     string
+		policy   *conf_v1.Policy
+		expected bool
+	}{
+		{
+			name:     "AccessControl is supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{AccessControl: &conf_v1.AccessControl{Allow: []string{"127.0.0.1"}}}},
+			expected: true,
+		},
+		{
+			name:     "CORS is supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{CORS: &conf_v1.CORS{}}},
+			expected: true,
+		},
+		{
+			name:     "IngressMTLS is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{IngressMTLS: &conf_v1.IngressMTLS{ClientCertSecret: "ca"}}},
+			expected: false,
+		},
+		{
+			name:     "EgressMTLS is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{EgressMTLS: &conf_v1.EgressMTLS{}}},
+			expected: false,
+		},
+		{
+			name:     "RateLimit is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{RateLimit: &conf_v1.RateLimit{}}},
+			expected: false,
+		},
+		{
+			name:     "JWTAuth is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{JWTAuth: &conf_v1.JWTAuth{}}},
+			expected: false,
+		},
+		{
+			name:     "BasicAuth is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{BasicAuth: &conf_v1.BasicAuth{}}},
+			expected: false,
+		},
+		{
+			name:     "OIDC is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{OIDC: &conf_v1.OIDC{}}},
+			expected: false,
+		},
+		{
+			name:     "APIKey is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{APIKey: &conf_v1.APIKey{}}},
+			expected: false,
+		},
+		{
+			name:     "WAF is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{WAF: &conf_v1.WAF{}}},
+			expected: false,
+		},
+		{
+			name:     "Cache is not supported",
+			policy:   &conf_v1.Policy{Spec: conf_v1.PolicySpec{Cache: &conf_v1.Cache{}}},
+			expected: false,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			if got := IsPolicySupportedOnIngress(tc.policy); got != tc.expected {
+				t.Errorf("IsPolicySupportedOnIngress() = %v, want %v", got, tc.expected)
+			}
+		})
+	}
+}
+
+func TestGeneratePolicies_UnsupportedOnIngress(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	ingOwnerDetails := policyOwnerDetails{
+		owner:           nil,
+		ownerName:       "cafe-ingress",
+		ownerNamespace:  "default",
+		parentNamespace: "default",
+		parentName:      "cafe-ingress",
+		parentType:      "ing",
+	}
+	policyRef := []conf_v1.PolicyReference{{Name: "test-policy", Namespace: "default"}}
+
+	unsupportedPolicies := map[string]*conf_v1.Policy{
+		"IngressMTLS": {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{IngressMTLS: &conf_v1.IngressMTLS{ClientCertSecret: "ca"}}},
+		"EgressMTLS":  {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{EgressMTLS: &conf_v1.EgressMTLS{}}},
+		"RateLimit":   {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{RateLimit: &conf_v1.RateLimit{}}},
+		"JWTAuth":     {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{JWTAuth: &conf_v1.JWTAuth{}}},
+		"BasicAuth":   {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{BasicAuth: &conf_v1.BasicAuth{}}},
+		"OIDC":        {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{OIDC: &conf_v1.OIDC{}}},
+		"APIKey":      {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{APIKey: &conf_v1.APIKey{}}},
+		"WAF":         {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{WAF: &conf_v1.WAF{}}},
+		"Cache":       {ObjectMeta: meta_v1.ObjectMeta{Name: "test-policy", Namespace: "default"}, Spec: conf_v1.PolicySpec{Cache: &conf_v1.Cache{}}},
+	}
+
+	for policyType, pol := range unsupportedPolicies {
+		t.Run(policyType, func(t *testing.T) {
+			t.Parallel()
+			policies := map[string]*conf_v1.Policy{"default/test-policy": pol}
+			result, warnings := generatePolicies(ctx, ingOwnerDetails, policyRef, policies, "spec", "", policyOptions{}, nil)
+
+			if result.ErrorReturn == nil || result.ErrorReturn.Code != 500 {
+				t.Errorf("generatePolicies() expected ErrorReturn{500} for unsupported policy type %s on Ingress, got %v", policyType, result.ErrorReturn)
+			}
+			if len(warnings[nil]) == 0 {
+				t.Errorf("generatePolicies() expected a warning for unsupported policy type %s on Ingress", policyType)
+			}
+		})
+	}
+}
+
 func TestGenerateLRZPolicyGroupMap(t *testing.T) {
 	t.Parallel()
 	tests := []struct {

--- a/internal/k8s/policy.go
+++ b/internal/k8s/policy.go
@@ -135,14 +135,12 @@ func (lbc *LoadBalancerController) syncPolicy(task task) {
 
 	// Loop through the resources that reference this policy and check if the policy type is supported on the resource. If not, log an error and emit an event.
 	// Note: if we ever support all policy types on all resources, this loop can be removed.
-	var filteredResources []Resource
 	for _, res := range resources {
 		switch impl := res.(type) {
 		// We only check for Ingress resources because VirtualServer and VirtualServerRoute support all policy types.
 		//   If a new resource type is added that supports a subset of policy types, a new case should be added here to check for supported policy types on that resource.
 		case *IngressConfiguration:
 			if !polExists {
-				filteredResources = append(filteredResources, res)
 				continue
 			}
 			pol := obj.(*conf_v1.Policy)
@@ -151,15 +149,13 @@ func (lbc *LoadBalancerController) syncPolicy(task task) {
 					pol.Namespace, pol.Name, impl.Ingress.Namespace, impl.Ingress.Name)
 				nl.Error(lbc.Logger, msg)
 				lbc.recorder.Event(impl.Ingress, api_v1.EventTypeWarning, nl.EventReasonRejected, msg)
-				// Do not add to filteredResources: skip reloading this Ingress for an unsupported policy.
-				continue
+				// The reload still proceeds so that generatePolicies() surfaces the error
+				// (ErrorReturn 500) consistently regardless of which path triggered the sync.
 			}
-			filteredResources = append(filteredResources, res)
 		default:
-			filteredResources = append(filteredResources, res)
+			continue
 		}
 	}
-	resources = filteredResources
 
 	resourceExes := lbc.createExtendedResources(resources)
 

--- a/internal/k8s/policy.go
+++ b/internal/k8s/policy.go
@@ -138,7 +138,7 @@ func (lbc *LoadBalancerController) syncPolicy(task task) {
 	for _, res := range resources {
 		switch impl := res.(type) {
 		// We only check for Ingress resources because VirtualServer and VirtualServerRoute support all policy types.
-		//   If a new resource type is added that supports a subset of policy types, a new case should be added here to check for supported policy types on that resource.
+		// If Ingress support for a policy type is added in the future, the policy Spec must also be added in IsPolicySupportedOnIngress() in internal/configs/policy.go.
 		case *IngressConfiguration:
 			if !polExists {
 				continue

--- a/internal/k8s/policy.go
+++ b/internal/k8s/policy.go
@@ -135,44 +135,31 @@ func (lbc *LoadBalancerController) syncPolicy(task task) {
 
 	// Loop through the resources that reference this policy and check if the policy type is supported on the resource. If not, log an error and emit an event.
 	// Note: if we ever support all policy types on all resources, this loop can be removed.
+	var filteredResources []Resource
 	for _, res := range resources {
 		switch impl := res.(type) {
 		// We only check for Ingress resources because VirtualServer and VirtualServerRoute support all policy types.
 		//   If a new resource type is added that supports a subset of policy types, a new case should be added here to check for supported policy types on that resource.
 		case *IngressConfiguration:
 			if !polExists {
+				filteredResources = append(filteredResources, res)
 				continue
 			}
 			pol := obj.(*conf_v1.Policy)
-			switch {
-			case pol.Spec.CORS != nil:
-				// CORS policy is supported on Ingress
-				continue
-			case pol.Spec.AccessControl != nil:
-				// Access Control policy is supported on Ingress
-				continue
-			case pol.Spec.WAF != nil:
-				// WAF policy is supported on Ingress
-				continue
-			case pol.Spec.ExternalAuth != nil:
-				// External Auth policy is supported on Ingress
-				continue
-			case pol.Spec.IngressMTLS != nil:
-				// IngressMTLS policy is supported on Ingress
-				continue
-			case pol.Spec.EgressMTLS != nil:
-				// Egress MTLS policy is supported on Ingress
-				continue
-			default: // Unsupported policy type on Ingress
+			if !configs.IsPolicySupportedOnIngress(pol) {
 				msg := fmt.Sprintf("Policy %s/%s has unsupported type on Ingress resource %s/%s",
 					pol.Namespace, pol.Name, impl.Ingress.Namespace, impl.Ingress.Name)
 				nl.Error(lbc.Logger, msg)
 				lbc.recorder.Event(impl.Ingress, api_v1.EventTypeWarning, nl.EventReasonRejected, msg)
+				// Do not add to filteredResources: skip reloading this Ingress for an unsupported policy.
+				continue
 			}
+			filteredResources = append(filteredResources, res)
 		default:
-			continue
+			filteredResources = append(filteredResources, res)
 		}
 	}
+	resources = filteredResources
 
 	resourceExes := lbc.createExtendedResources(resources)
 

